### ISSUE TITLE
Restore back API performance enhancements for orders reverted in 28745

### ISSUE THREAD

**Issue by peterfabian** [woothemes/woocommerce#30981]
In #28745 we reverted some of the performance enhancements we were planning to add in 4.9 because of a bug that we found during the release process.
This issue scopes re-implementing those enhancements for orders, by making sure that wc_get_orders and wc_get_products support WP_Query filter parameters, along with adding tests for all those params.

**Comment by peterfabian** [comment]
Similar to #29184, just for orders.

**Comment by jeffstieler** [comment]
I've encountered what I believe to be a potential issue with maintaining backwards compatibility when reintroducing these changes.

Swapping to `wc_get_orders()` from `WP_Query` (internally to `WC_REST_CRUD_Controller`) will break queries for stores using the [`woocommerce_rest_orders_prepare_object_query`](https://github.com/woocommerce/woocommerce/blob/3611d4643791bad87a0d3e6e73e031bb80447417/plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-orders-v2-controller.php#L524) filter to add `meta_query` values. (`WC_Data_Store_WP::get_wp_query_args()` [strips it out](https://github.com/woocommerce/woocommerce/blob/3611d4643791bad87a0d3e6e73e031bb80447417/plugins/woocommerce/includes/data-stores/class-wc-data-store-wp.php#L276))

Maintaining backwards compatibility here isn't straightforward and might even require filtering the provided values back in using [`woocommerce_order_data_store_cpt_get_orders_query`](https://github.com/woocommerce/woocommerce/blob/3611d4643791bad87a0d3e6e73e031bb80447417/plugins/woocommerce/includes/data-stores/class-wc-order-data-store-cpt.php#L847).

An alternative would be to remove support for `meta_query` additions through `woocommerce_rest_orders_prepare_object_query` on a new `/v4/` endpoint.

What's interesting is that `get_wp_query_args()` initially supported `meta_query` but it was [removed during PR review](https://github.com/woocommerce/woocommerce/commit/6aecb15505ed0e5c6c4c97c652e37b967a050a68#diff-e0431622838bf76a4635ab1862fcb83bc6e9d3b09bc78831fd79b17470f1e3e6).

I'm soliciting feedback here on the best path forward. Thanks.

**Comment by peterfabian** [comment]
cc @vedanshujain for potential comments.

**Comment by vedanshujain** [comment]
Linking https://github.com/woocommerce/woocommerce/pull/31222 where we are discussing this change.

**Comment by lsinger** [comment]
Is this a duplicate of #29184?

**Comment by coreymckrill** [comment]
^ Not a duplicate, this issue is for orders, the other one is for products.

**Comment by samrat-biswas-git** [comment]
This needs engineering input before we can confirm the final disposition.

Area to validate: The issue maps to a real historical revert, and trunk shows a split world: HPOS order queries already support the newer path, while the CPT REST controller still avoids it. Because the unresolved question is compatibility and architecture for legacy CPT stores rather than simple missing code, engineering judgment is needed before this becomes ready implementation work.
